src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,18 +2205,28 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
+                var enumerator = new Enumerator(result);
+                try
                 {
-                    if (match(item))
+                    var startIndex = 0;
+                    while(enumerator.MoveNext())
                     {
-                        result = result.RemoveAt(index);
-                    }
-                    else
-                    {
-                        index++;
+                        if (match(enumerator.Current))
+                        {
+                            result = result.RemoveAt(startIndex);
+                            enumerator.Dispose();
```

Comment:
ok... I had that at one point and removed it as it felt unnecessary, given the rule that calling Dispose twice should never have adverse impact (see https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern)
```✓ DO allow the Dispose(bool) method to be called more than once. The method might choose to do nothing after the first call.```

If you prefer we clear it, we can do so.  I can also look into removing the likelihood of the new stack allocation (though it seems that shouldn't be a problem since we're just putting the object back into the pool and pulling it back out ;)